### PR TITLE
Login networking

### PR DESCRIPTION
**Pull request by guvanch-amangeldiyev**

### CODE REVIEW

File: packages/presentation/lib/screen/login/login_view/main_login.dart [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    bloc.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return StreamPlatformStackContent(
+        dataStream: bloc.dataStream,
+        children: (blocData) {
+          //final screenData = blocData.data;
```
Comment (ilya-safyan-idf) [inline packages/presentation/lib/screen/login/login_view/main_login.dart]: delete unnecessary comments

Comment (guvanch-amangeldiyev) [reply packages/presentation/lib/screen/login/login_view/main_login.dart]: fixed

**Review by ilya-safyan-idf — APPROVED** [review]

File: .gitignore [hunk 2]
```diff
@@ -44,7 +47,7 @@ app.*.map.json
...
 /android/app/profile
 /android/app/release
-/packages/domain/pubspec.lock
-/packages/presentation/pubspec.lock
-/packages/data/pubspec.lock
-/pubspec.lock
+packages/presentation/pubspec.lock
+packages/data/pubspec.lock
+pubspec.lock
+packages/domain/pubspec.lock
```
Comment (sergei-mikhailovskii-idf) [inline .gitignore]: ignore all pubspec.lock in one line, like **/pubspec.lock

File: .gitignore [hunk 3]
```diff
@@ -30,6 +30,9 @@
 .pub-cache/
 .pub/
 /build/
+pubspec.lock
```
Comment (sergei-mikhailovskii-idf) [inline .gitignore:34]: why do u ignore pubspec.lock twice? here and lower

File: packages/data/lib/di/data_injector.dart [hunk 4]
```diff
@@ -7,34 +7,29 @@ import 'package:domain/repository/network_repository.dart';
...
-  _initApiModule();
-}
+  final sl = GetIt.I;
 
-void _initApiModule() {
-  GetIt.I.registerFactory<CancelToken>(() => CancelToken());
-  _initDioModule();
-  _initServiceModule();
-}
+  sl.registerSingleton<CancelToken>(
```
Comment (sergei-mikhailovskii-idf) [inline packages/data/lib/di/data_injector.dart]: singleton -> factory

File: packages/domain/lib/usecase/login_usecase.dart [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  LoginStepUseCase(
+    this._repository,
+    this._cache,
+  );
+
+  @override
+  Future<void> call(LoginParms parms) async =>
+      _repository.getJobs().then((value) {
+        _cache.jobs =
+            value!.data?.map((element) => Job.fromJson(element)).toList();
```
Comment (sergei-mikhailovskii-idf) [inline packages/domain/lib/usecase/login_usecase.dart]: ! -> ?

File: packages/presentation/build/.last_build_id [hunk 6]
```diff
@@ -0,0 +1 @@
+c527f3a774fe3cebd1e7d965a7ec8bc9
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/build/.last_build_id]: ignore all build files

File: packages/presentation/lib/l10n/app_en.arb [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+{
+  "appTitle": "App Title",
+  "textWithParam": "Text with {param}",
+  "@textWithParam": {
+    "placeholders": {
+      "param": {
+        "type": "String"
+      }
+    }
+  }
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/l10n/app_en.arb]: why do u need it?

File: packages/presentation/lib/l10n/app_ru.arb [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+{
+  "appTitle": "Название приложения",
+  "textWithParam": "Текст с {param}",
+  "@textWithParam": {
+    "placeholders": {
+      "param": {
+        "type": "String"
+      }
+    }
+  }
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/l10n/app_ru.arb]: -//-

File: packages/presentation/lib/screen/home/home_tab_bar.dart [hunk 9]
```diff
@@ -0,0 +1,41 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_svg/svg.dart';
+import 'package:presentation/core/theme/theme_app.dart';
+
+const List<Tab> _myMainTabs = <Tab>[
+  Tab(
+    text: "Card ES",
+  ),
+  Tab(text: "Card MX"),
+  Tab(text: "MMES"),
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/screen/home/home_tab_bar.dart]: tabs should be dynamic - ask Andrey, he'll explain u where u should take them

File: packages/presentation/lib/screen/home/home_tab_bar.dart [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  @override
+  Widget build(BuildContext context) {
+    return DefaultTabController(
+      length: _myMainTabs.length,
+      child: Scaffold(
+        body: const TabBarView(children: _myWidgets),
+        appBar: AppBar(
+          backgroundColor: AppColors.accentGreen,
+          bottom: const TabBar(
+              labelStyle: TextStyle(fontSize: 10), tabs: _myMainTabs),
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/screen/home/home_tab_bar.dart:35]: comma

File: packages/presentation/lib/screen/login/bloc/login_bloc.dart [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+          log("sucsecc");
+          //! trigger of navigate
+        }
+      },
+      errorAction: (e) {
+        if (e is AuthException) {
+          _screenData.exception = e;
+        }
+      },
+    );
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/screen/login/bloc/login_bloc.dart]: should be refactored. ask Andrey/Ilya/Alexei, they can help u

File: packages/presentation/lib/screen/login/bloc/login_bloc.dart [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+  @override
+  String? validateTextEmail(String formEmail) {
+    if (formEmail.isEmpty) return "E-mail is requried.";
+    // String pattern = r'\w+@\W+\.\w+';
+    // RegExp reg = RegExp(pattern);
+    if (!formEmail.contains("@")) {
+      return ErrorTextField.login_invalid;
+    }
+    return null;
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/screen/login/bloc/login_bloc.dart]: ?

File: packages/presentation/lib/screen/login/bloc/login_bloc.dart [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  @override
+  String? validateTextPassword(String formPassword) {
+    if (formPassword.isEmpty) return "password is requried.";
+
+    if (formPassword.length < 6) {
+      return ErrorTextField.password_invalid;
+    }
+    return null;
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/screen/login/bloc/login_bloc.dart]: ?

File: packages/presentation/lib/screen/login/login_view/login_key.dart [hunk 14]
```diff
@@ -0,0 +1,59 @@
+import 'package:flutter/material.dart';
+import 'package:presentation/core/helpers/primary_button.dart';
+import 'package:presentation/widgets/text_field.dart';
+
+const String _login = "Login";
+const String _password = "Password";
+
+class LoginPageWithKey extends StatelessWidget {
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/screen/login/login_view/login_key.dart]: why did u extract it to a separate widget? it's ok to keep it in LoginPage

File: packages/presentation/lib/screen/login/login_view/login_key.dart [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+  Widget build(BuildContext context) {
+    return Scaffold(
+        body: SafeArea(
+      child: Form(
+        key: keyStore,
+        child: Column(
+          children: [
+            const SizedBox(height: 20),
+            LoginTextField(
+              onSaved: (value) => textField[_login] = value ?? "",
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/screen/login/login_view/login_key.dart]: why do u save the value in onSaved, but not onChange?

File: packages/presentation/lib/screen/login/login_view/login_key.dart [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+              obscure: false,
+              validator: validateEmail,
+              //onChanged: (String value) {},
+              text: _login,
+            ),
+            const SizedBox(height: 20),
+            LoginTextField(
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/screen/login/login_view/login_key.dart]: what's the difference between LoginTextField and any other TextFormField in the app? why did u extract it to separate widget?

File: packages/presentation/lib/screen/login/login_view/main_login.dart [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+        children: (blocData) {
+          return LoginPageWithKey(
+            validateEmail: (value) => bloc.validateTextEmail(value ?? ''),
+            validatePassword: (value) => bloc.validateTextPassword(value ?? ''),
+            onChangedEmail: bloc.setLogin,
+            onChangedPassword: bloc.setPassword,
+            onPressLogin: bloc.login,
+            keyStore: bloc.keyStore,
+            textField: bloc.onSave,
+          );
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/screen/login/login_view/main_login.dart]: no need in separating fields. move them here

**Review by sergei-mikhailovskii-idf — CHANGES_REQUESTED** [review]

File: packages/presentation/lib/screen/login/login_view/login_key.dart (hunk 14 above)
Comment (guvanch-amangeldiyev) [reply packages/presentation/lib/screen/login/login_view/login_key.dart]: by separating ui would be cleaner

File: .gitignore (hunk 2 above)
Comment (guvanch-amangeldiyev) [reply .gitignore]: ignored

File: .gitignore (hunk 3 above)
Comment (guvanch-amangeldiyev) [reply .gitignore:34]: I didn't see the the lower(, ignored

File: packages/domain/lib/usecase/login_usecase.dart (hunk 5 above)
Comment (guvanch-amangeldiyev) [reply packages/domain/lib/usecase/login_usecase.dart]: also changed

File: packages/presentation/build/.last_build_id (hunk 6 above)
Comment (guvanch-amangeldiyev) [reply packages/presentation/build/.last_build_id]: ignored

File: packages/presentation/lib/screen/login/bloc/login_bloc.dart (hunk 13 above)
Comment (guvanch-amangeldiyev) [reply packages/presentation/lib/screen/login/bloc/login_bloc.dart]: fixed

File: packages/presentation/lib/screen/login/login_view/login_key.dart (hunk 14 above)
Comment (guvanch-amangeldiyev) [reply packages/presentation/lib/screen/login/login_view/login_key.dart]: I read the extracting to methods is bad for performance that way I extract to widgets

File: packages/presentation/lib/screen/login/login_view/login_key.dart (hunk 15 above)
Comment (guvanch-amangeldiyev) [reply packages/presentation/lib/screen/login/login_view/login_key.dart]: onChanged was committed, fixed

File: packages/presentation/lib/screen/login/login_view/main_login.dart (hunk 17 above)
Comment (guvanch-amangeldiyev) [reply packages/presentation/lib/screen/login/login_view/main_login.dart]: fixed

File: packages/presentation/lib/l10n/app_en.arb (hunk 7 above)
Comment (guvanch-amangeldiyev) [reply packages/presentation/lib/l10n/app_en.arb]: I use them just for test when set up localization

File: packages/presentation/lib/screen/login/bloc/login_bloc.dart [hunk 18]
```diff
@@ -27,68 +38,64 @@ abstract class LoginBloc implements BaseBloc {
...
+    super.handleData(
+      isLoading: isLoading,
+      data: _screenData,
+    );
   }
 
   @override
-  void setPassword(String password) {
-    _screenData.passwordInput = password;
+  void navigateToHomePage() {
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/screen/login/bloc/login_bloc.dart]: private

File: packages/presentation/lib/screen/login/bloc/login_bloc.dart [hunk 19]
```diff
@@ -97,37 +104,57 @@ class _LoginBloc extends BlocImpl implements LoginBloc {
...
-
         _updateData();
 
         navigateToHomePage();
       },
       errorAction: (e) {
         if (e is AuthException) {
           _screenData.exception = e;
+          if (loginFieldKey.currentState != null &&
+              !loginFieldKey.currentState!.validate()) {
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/screen/login/bloc/login_bloc.dart]: loginFieldKey.currentState?.validate() != true

File: packages/presentation/lib/screen/login/bloc/login_bloc.dart [hunk 20]
```diff
@@ -97,37 +104,57 @@ class _LoginBloc extends BlocImpl implements LoginBloc {
...
+            loginFocusNode.requestFocus();
+          }
+          if (passwordFieldKey.currentState != null &&
+              !passwordFieldKey.currentState!.validate()) {
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/screen/login/bloc/login_bloc.dart]: passwordFieldKey.currentState!.validate() != true

**Review by sergei-mikhailovskii-idf — APPROVED** [review]

**Pull request merged**